Docs overhaul (#431)
cmu-delphi/epipredict: vignettes/custom_epiworkflows.Rmd

---
title: "Custom Epiworkflows"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Custom Epiworkflows}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
source(here::here("vignettes/_common.R"))
```

```{r setup, message=FALSE, include = FALSE}
library(dplyr)
library(parsnip)
library(workflows)
library(recipes)
library(epipredict)
library(epiprocess)
library(ggplot2)
forecast_date <- as.Date("2021-08-01")
used_locations <- c("ca", "ma", "ny", "tx")
library(epidatr)
```

> To get a better handle on custom `epi_workflow()`s, lets recreate and then

Author: yeah that makes sense. In addition to what you wrote, and fixing a typo, I'm going to switch it to
```
If you want to do custom data preprocessing or fit a model that isn't included in the canned workflows, you'll need to write a custom `epi_workflow()`.
An `epi_workflow()` is a sub-class of a `workflows::workflow()` from the
`{workflows}` package designed to handle panel data specifically.

To understand how to work with custom `epi_workflow()`s, let's recreate and then
```

    - undo operations done in the steps, such as convert back to counts from rates
    - threshold so the forecast doesn't include negative values
    - generally adapt the format of the prediction to it's eventual use.

# Recreating `four_week_ahead` in an `epi_workflow()`
To be able to extend this beyond what `arx_forecaster()` itself will let us do,
we first need to understand how to recreate it using a custom `epi_workflow()`.

To use a custom workflow, there are a couple of steps:

1. define the `epi_recipe()`, which contains the preprocessing steps
2. define the `frosting()` which contains the post-processing layers
3. Combine these with a trainer such as `quantile_reg()` into an
   `epi_workflow()`, which we can then fit on the training data
4. `fit()` the workflow on some data
5. grab the right prediction data using `get_test_data()` and apply the fit data
   to generate a prediction

## Define the `epi_recipe()`
To do this, we'll first take a look at the steps as they're found in
`four_week_ahead`:

```{r inspect_fwa_steps, warning=FALSE}
hardhat::extract_recipe(four_week_ahead$epi_workflow)
```

So there are 6 steps we will need to recreate.

> One thing to note about the extracted recipe is that it has already been

Reviewer: I don't think we talk explicitly about what recipes are. Let me make sure I understand and I'll make a commit.

So the workflow is the entire procedure (preprocessor + trainer + postprocessor). The preprocessor is also called a recipe; it is not optional. The trainer is also called a model or engine (are these the same thing?); it is not optional. The postprocessor is  also called frosting; it is optional.

> One thing to note about the extracted recipe is that it has already been

Author: that's right yeah.

> One thing to note about the extracted recipe is that it has already been

Reviewer: Okay, I'll add a blurb about this.

dataset on which you are going to train. 
However, it should have the same columns and the same metadata, such as `as_of`
or `other_keys`; it is typically easiest to just use the training data itself.

Then we add each step via pipes; in principle the order matters, though for this
recipe only `step_epi_naomit()` and `step_training_window()` depend on the steps
before them:

```{r make_steps}
four_week_recipe <- four_week_recipe |>
  step_epi_lag(case_rate, lag = c(0, 1, 2, 3, 7, 14)) |>
  step_epi_lag(death_rate, lag = c(0, 7, 14)) |>
  step_epi_ahead(death_rate, ahead = 4 * 7) |>
  step_epi_naomit() |>
  step_training_window()
```

One thing to note: we only added 5 steps here because `step_epi_naomit()` is
actually a wrapper around adding 2 base `step_naomit()`s, one for
`all_predictors()` and one for `all_outcomes()`, differing in their treatment at
predict time.

`step_epi_lag()` and `step_epi_ahead()` both accept tidy syntax, so if for
example we wanted the same lags for both `case_rate` and `death_rate`, we could
have done `step_epi_lag(ends_with("rate"), lag = c(0, 7, 14))`.

In general for the `{recipes}` package, steps assign roles, such as `predictor`

> and `outcome` to columns, either by adding new columns or adjusting existing

Reviewer 2: Personal preference would be for "at least slightly self-contained". That is, you should probably provide enough info that a typical user can understand it just from reading here. Then link there for details.

`recipe`[^2]:

```{r make_frosting}
four_week_layers <- frosting() |>
  layer_predict() |>
  layer_residual_quantiles(quantile_levels = c(0.1, 0.25, 0.5, 0.75, 0.9)) |>
  layer_add_forecast_date() |>
  layer_add_target_date() |>
  layer_threshold()
```

Most layers will work for any engine or steps; `layer_predict()` you will want
to call in every case.
There are a couple of layers, however, which depend on whether the engine used predicts quantiles or point estimates.

The layers that are only supported by point estimate engines (such as
`linear_reg()`) are

- `layer_residual_quantiles()`: the preferred method of generating quantiles for
  non-quantile models, it uses the error residuals of the engine. 
  This will work for most parsnip engines.
- `layer_predictive_distn()`: alternate method of generating quantiles, it uses
  an approximate parametric distribution. This will work for linear regression
  specifically.
  
TODO check this
  
On the other hand, the layers that are only supported by quantile estimating
engines (such as `quantile_reg()`) are

- `layer_quantile_distn()`: adds the specified quantiles.

>   If they differ from the ones actually fit, they will be interpolated and/or

Author: ```suggestion
  If the quantile levels specified differ from the ones actually fit, they will be interpolated and/or extrapolated.
```

>   If they differ from the ones actually fit, they will be interpolated and/or

Reviewer: https://github.com/cmu-delphi/epipredict/pull/431/commits/00f8ce524b4e0170eac022ba28ee24c300e4bebb

  extrapolated.
- `layer_point_from_distn()`: this adds the median quantile as a point estimate,
  and should be called after `layer_quantile_distn()` if called at all. 

## Fitting an `epi_workflow()`

Given that we now have a recipe and some layers, we need to assemble the workflow:
```{r workflow_building}
four_week_workflow <- epi_workflow(
  four_week_recipe,
  linear_reg(),
  four_week_layers
)
```

And fit it to recreate `four_week_ahead$epi_workflow`
```{r workflow_fitting}
fit_workflow <- four_week_workflow |> fit(training_data)
```

## Predicting

> To do a prediction, we need to first narrow the dataset down to the relevant

Author: It's just the fact that we're predicting on all values. I say this later in the section, but it probably makes sense to lead with it.

> To do a prediction, we need to first narrow the dataset down to the relevant

Author: Added a re-wording as a suggestion, but it won't show up here b/c this is out-dated. Going to mark this as resolved and suggest further discussion there.

datapoints:

```{r grab_data}
relevant_data <- get_test_data(
  four_week_recipe,
  training_data
)
```

With a fit workflow and test data in hand, we can actually make our predictions:

```{r workflow_pred}
fit_workflow |> predict(relevant_data)
```

Note that if we had simply plugged `training_data` into `predict()` we still get
predictions:

```{r workflow_pred_training}
fit_workflow |> predict(training_data)
```

The resulting tibble is 800 rows long, however.

> This produces forecasts for not just the actual `forecast_date`, but for every

Reviewer: question: where did we set the `forecast_date`/how does the workflow know what it is? what if we want to use a different forecast date? do we have to re-define and re-compile the whole workflow?

Author: This is set in `epi_recipe`. I'll make some changes to account for this. Also I'll have to change the name to `reference_date`, since that's what's in that function.

day in the dataset it has enough data to actually make a prediction.
To narrow this down, we could filter to rows where the `time_value` matches the `forecast_date`:

```{r workflow_pred_training_filter}
fit_workflow |>
  predict(training_data) |>
  filter(time_value == forecast_date)
```

This can be useful for cases where `get_test_data()` doesn't pull sufficient

> data.

Reviewer: issue: This sounds suspicious. If `get_test_data` decided there wasn't sufficient data, it sounds like the `predict` -> `filter` approach is doing something wrong (predicting with insufficient data).

Reviewer: I assumed that the `predict` -> `filter` approach would return the same predictions for `forecast_date`, but it sounds like not.

Author: It does produce the same forecasts when it produces forecasts, it's just that sometimes `get_test_data` won't produce a forecast. `get_test_data` is just dumb and won't necessarily grab enough data if the recipe is sufficiently complicated. It's a hack that we need to rework.

Reviewer: Modified slightly in https://github.com/cmu-delphi/epipredict/pull/431/commits/d218ed696cf67ceef012301fbfb905cd9d68d8cb

preferred prediction format.
Another would be to include a time component to the prediction (useful if we
expect there to be a strong seasonal component).

## Growth rate

One feature that may potentially improve our forecast is looking at the growth
rate

```{r growth_rate_recipe}
growth_rate_recipe <- epi_recipe(
  covid_case_death_rates |>
    filter(time_value <= forecast_date, geo_value %in% used_locations)
) |>
  step_epi_lag(case_rate, lag = c(0, 1, 2, 3, 7, 14)) |>
  step_epi_lag(death_rate, lag = c(0, 7, 14)) |>
  step_epi_ahead(death_rate, ahead = 4 * 7) |>
  step_epi_naomit() |>
  step_growth_rate(death_rate) |>
  step_training_window()
```

Inspecting the newly added column:

```{r growth_rate_print}
growth_rate_recipe |>
  prep(training_data) |>
  bake(training_data) |>
  select(
    geo_value, time_value, case_rate,
    death_rate, gr_7_rel_change_death_rate
  )
```

And the role:

```{r growth_rate_roles}
prepped <- growth_rate_recipe |>
  prep(training_data)
prepped$term_info |> filter(grepl("gr", variable))
```

To demonstrate the changes in the layers that come along with it, we will use

> `quantile_reg()` as the model, which requires changing from `layer_residual_quantiles()` to `layer_quantile_distn()` and `layer_point_from_distn()`:

Author: yeah this is probably not the best place to demo this, it's not necessary. I talk about it briefly in [Define the Frosting](https://67e478bce7f35616d80fb137--epipredict.netlify.app/articles/custom_epiworkflows#define-the-frosting). I should probably just add a separate section with an example of changing the engine in a custom workflow (and we should probably just abstract these layers out with a layer that auto-manages this!).

The short reason is we produce quantiles and point predictions using layers, and `linear_reg()` produces point+residuals, whereas `quantile_reg` produces quantiles.

> `quantile_reg()` as the model, which requires changing from `layer_residual_quantiles()` to `layer_quantile_distn()` and `layer_point_from_distn()`:

Reviewer: Clarified in https://github.com/cmu-delphi/epipredict/pull/431/commits/23baa026500b8392a8ea79233cbebae4df903d20

    dates = rep(forecast_date - 7 * 2, 2 * length(used_locations)),
    heights = c(rep(150, 4), rep(0.30, 4))
  )

result_plot <- autoplot(
  object = gr_fit_workflow,
  predictions = gr_predictions,
  plot_data = covid_case_death_rates |>
    filter(geo_value %in% used_locations, time_value > "2021-07-01")
) +
  geom_vline(aes(xintercept = forecast_date)) +
  geom_text(
    data = forecast_date_label %>% filter(.response_name == "death_rate"),
    aes(x = dates, label = "forecast\ndate", y = heights),
    size = 3, hjust = "right"
  ) +
  scale_x_date(date_breaks = "3 months", date_labels = "%Y %b") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```
</details>
```{r, echo=FALSE}
result_plot
```

TODO `get_test_data` isn't actually working here...

## Population scaling

Suppose we're sending our predictions to someone who is looking to understand
counts, rather than rates.
Then we can adjust just the frosting to get a forecast for the counts from our
rates forecaster:

```{r rate_scale}
count_layers <-
  frosting() |>
  layer_predict() |>
  layer_residual_quantiles(quantile_levels = c(0.1, 0.25, 0.5, 0.75, 0.9)) |>

>   layer_population_scaling(

Author: These are all good questions. Partly they're passed separately b/c communicating between steps and layers is a bit of a pain. But https://github.com/cmu-delphi/epipredict/pull/451 is a template for how we could rewrite this to auto-pull the value from the relevant step. In general it badly needs an overhaul, probably including default scalings based on the epi_df's `geo_type` (with this sort of user-specified behavior being a fall-back).

count_predictions
```

which are 2-3 orders of magnitude larger than the corresponding rates above.
`df` represents the scaling value; in this case it is the state populations,
while `rate_rescaling` gives the denominator of the rate (our fit values were
per 100,000).

# Custom classifier workflow

As a more complicated example of the kind of pipeline that you can build using
this framework, here is an example of a hotspot prediction model, which predicts
whether the case rates are increasing (`up`), decreasing (`down`) or flat
(`flat`).
This comes from a paper by McDonald, Bien, Green, Hu et al[^3], and roughly
serves as an extension of `arx_classifier()`.

First, we need to add a factor version of the `geo_value`, so that it can
actually be used as a feature.

```{r training_factor}
training_data <-
  training_data %>%
  mutate(geo_value_factor = as.factor(geo_value))
```

Then we put together the recipe, using a combination of base `{recipe}`
functions such as `add_role()` and `step_dummy()`, and `{epipreict}` functions
such as `step_growth_rate()`.

```{r class_recipe}
classifier_recipe <- epi_recipe(training_data) %>%
  add_role(time_value, new_role = "predictor") %>%

>   step_dummy(geo_value_factor) %>%

Author: `step_dummy` makes a collection of indicator variables for each value, whereas `add_role` just assigns a role to an existing column.

>   step_dummy(geo_value_factor) %>%

Reviewer: https://github.com/cmu-delphi/epipredict/pull/431/commits/06df10dd2713250b5d48825e44edf3b8620d4cfd

$$
 Z_{\ell, t}=
    \begin{cases}
      \text{up}, & \text{if}\ Y^{\Delta}_{\ell, t} > 0.25 \\
      \text{down}, & \text{if}\  Y^{\Delta}_{\ell, t} < -0.20\\
      \text{flat}, & \text{otherwise}
    \end{cases}
$$

where $Y^{\Delta}_{\ell, t}$ is the growth rate at location $\ell$ and time $t$.
`up` means that the `case_rate` is has increased by at least 25%, while `down`
means it has decreased by at least 20%.

Note that both `step_growth_rate()` and `step_epi_ahead()` assign the role
`none` explicitly; this is because they're used as intermediate steps to create
both predictors and the outcome.
`step_rm()` drops them after they're done, along with the original `raw` columns
`death_rate` and `case_rate` (both `geo_value` and `time_value` are retained
because their roles have been reassigned).

To fit a classification model like this, we will need to use a parsnip model
with mode classification; the simplest example is `multinomial_reg()`.
We don't actually need to apply any layers, so we can skip adding one to the `epiworkflow()`:

```{r, warning=FALSE}
wf <- epi_workflow(
  classifier_recipe,
  multinom_reg()
) %>%
  fit(training_data)

> forecast(wf) %>% filter(!is.na(.pred_class))

Author: This is a section I copied pretty directly from a different vignette, so I didn't actually notice this. I'll rewrite to just use layers

  specifically.
  
<!-- TODO check this -->
  
On the other hand, the layers that are only supported by quantile estimating
engines (such as `quantile_reg()`) are

- `layer_quantile_distn()`: adds the specified quantiles.
  If they differ from the ones actually fit, they will be interpolated and/or
  extrapolated.
- `layer_point_from_distn()`: this adds the median quantile as a point estimate,
  and, if called, should be included after `layer_quantile_distn()`. 

## Fitting an `epi_workflow()`

Now that we have a recipe and some layers, we can assemble the workflow.
This is as simple as passing the component preprocessor, model, and postprocessor into `epi_workflow()`.

```{r workflow_building}
four_week_workflow <- epi_workflow(
  four_week_recipe,
  linear_reg(),
  four_week_layers
)
```

After fitting it, we will have recreated `four_week_ahead$epi_workflow`.

```{r workflow_fitting}
fit_workflow <- four_week_workflow |> fit(training_data)
```

Running `fit()` calculates all preprocessor-required parameters, and trains the model on the data passed in `fit()`.
However, it does not generate any predictions; predictions need to be created in a separate step.

## Predicting

> To make a prediction, we need to first narrow a data set down to the relevant observations.
> This process removes observations that will not be used for training because, for example, they contain missing values or <!-- TODO other reasons?-->.

Author: ```suggestion
To make a prediction, it helps to narrow the data set down to the relevant observations using `get_test_data()`. Not doing this will still fit, but it will predict on every day in the data-set, and not just on the `forecast_date`.
```

Here's a possible rephrasing

# Extending `four_week_ahead`

Now that we know how to create `four_week_ahead` from scratch, we can start modifying the workflow to get custom behavior.

There are many ways we could modify `four_week_ahead`. We might consider:

- Converting from rates to counts
- Including a growth rate estimate as a predictor
- Including a time component as a predictor -- useful if we
expect there to be a strong seasonal component to the outcome
- Scaling by a factor

We will demo a couple of these modifications below.

## Growth rate

Let's say we're interested in including growth rate as a predictor in our model because
we think it may potentially improve our forecast.
We can easily create a new growth rate column as a step in the `epi_recipe`.

```{r growth_rate_recipe}
growth_rate_recipe <- epi_recipe(
  covid_case_death_rates |>
    filter(time_value <= forecast_date, geo_value %in% used_locations)
) |>
  step_epi_lag(case_rate, lag = c(0, 1, 2, 3, 7, 14)) |>
  step_epi_lag(death_rate, lag = c(0, 7, 14)) |>
  step_epi_ahead(death_rate, ahead = 4 * 7) |>
  step_epi_naomit() |>
  # Calculate growth rate from death rate column.

>   step_growth_rate(death_rate) |>

Author: I guess it really ought to be happening at the top. `growth_rate` doesn't handle `na`'s all that gracefully; if you drop the `naomit` you get
```r
epi_recipe(filtered) |>
+   step_epi_lag(case_rate, lag = c(0, 1, 2, 3, 7, 14)) |>
+   step_epi_lag(death_rate, lag = c(0, 7, 14)) |>
+   step_epi_ahead(death_rate, ahead = 4 * 7) |>
+   step_growth_rate(death_rate) |>
+   # Calculate growth rate from death rate column.
+   prep(filtered) |> bake(filtered)
+ Error in `dplyr::transmute()`:
ℹ In argument: `across(...)`.
ℹ In group 1: `geo_value = "ca"`.
Caused by error in `across()`:
! Can't compute column `gr_7_rel_change_death_rate`.
Caused by error in `epiprocess::growth_rate()`:
! Either the first or last `y` values are not finite. This may be due to `log_scale =
  TRUE`.
```

>   step_growth_rate(death_rate) |>

Reviewer 2: That shouldn't be happening with default arguments. So this may be a bug or may be due to package version mismatch. Can you create an issue with a minimal reprex?

serves as an extension of `arx_classifier()`.

First, we need to add a factor version of `geo_value`, so that it can be used as a feature.

```{r training_factor}
training_data <-
  covid_case_death_rates |>
  filter(time_value <= forecast_date, geo_value %in% used_locations) |>
  mutate(geo_value_factor = as.factor(geo_value))
```

Then we put together the recipe, using a combination of base `{recipe}`
functions such as `add_role()` and `step_dummy()`, and `{epipredict}` functions
such as `step_growth_rate()`.

```{r class_recipe}
classifier_recipe <- epi_recipe(training_data) |>
  # Label `time_value` as predictor and do no other processing
  add_role(time_value, new_role = "predictor") |>
  # Use one-hot encoding on `geo_value_factor` and label each resulting column as a predictor
  step_dummy(geo_value_factor) |>
  # Create and lag `case_rate` growth rate
  step_growth_rate(case_rate, role = "none", prefix = "gr_") |>
  step_epi_lag(starts_with("gr_"), lag = c(0, 7, 14)) |>
  step_epi_ahead(starts_with("gr_"), ahead = 7, role = "none") |>
  # Divide growth rate into 3 bins, and label as outcome variable

>   # Note `recipes::step_cut()` has a bug that prevents us from using it here

Reviewer 2: This may be fixed. Can you check?

Author: It's kind of weird; if I just substitute it in, assigning roles via `step_cut(ahead_7_gr_7_rel_change_case_rate, breaks = c(-0.2, 0.25) / 7, role = "outcome")` doesn't actually work, but if I just let `step_epi_ahead` add the role, it works. `add_role` after doesn't seem to find the column for some mysterious reason. There's also no way to add labels to the values, so it's just predicts explicit ranges like `(0.0357, 0.262]` which I guess works.

Reviewer 2: Looking at the code, it seems to always create closed ranges. I actually suspect this bug may have gone away, but I can't tell. Let's remove the note and add an issue to try to track it later.

- growth rate of case rate, both at prediction time (no lag), and lagged by one and two weeks

The outcome variable is created by composing several steps together. `step_epi_ahead()`
creates a column with the growth rate one week into the future, and
`step_mutate()` turns that column into a factor with 3 possible values,

$$
 Z_{\ell, t}=
    \begin{cases}
      \text{up}, & \text{if}\ Y^{\Delta}_{\ell, t} > 0.25 \\
      \text{down}, & \text{if}\  Y^{\Delta}_{\ell, t} < -0.20\\
      \text{flat}, & \text{otherwise}
    \end{cases}
$$

where $Y^{\Delta}_{\ell, t}$ is the growth rate at location $\ell$ and time $t$.
`up` means that the `case_rate` is has increased by at least 25%, while `down`
means it has decreased by at least 20%.

Note that in both `step_growth_rate()` and `step_epi_ahead()` we explicitly assign the role
`none`. This is because those columns are used as intermediaries to create
predictor and outcome columns.
Afterwards, `step_rm()` drops the temporary columns, along with the original `role = "raw"` columns
`death_rate` and `case_rate`. Both `geo_value_factor` and `time_value` are retained
because their roles have been reassigned.

> To fit a classification model like this, we will need to use a `{parsnip}` model

Reviewer 2: Maybe state that this is 3-class classification

that has `mode = "classification"`.
The simplest example of a `{parsnip}` `classification`-`mode` model is `multinomial_reg()`.
The needed layers are more or less the same as the `linear_reg()` regression layers, with the addition that we need to remove some `NA` values:

```{r, warning=FALSE}
frost <- frosting() |>
  layer_naomit(starts_with(".pred")) |>
  layer_add_forecast_date() |>
  layer_add_target_date() |>

>   layer_threshold()

Reviewer 2: We don't need to threshold do we? Actually, this should probably be an error.
Let's add an issue.

Author: ha, yeah that doesn't make any sense. removed and filed https://github.com/cmu-delphi/epipredict/issues/465

```

```{r, warning=FALSE}
wf <- epi_workflow(
  classifier_recipe,
  multinom_reg(),
  frost
) |>
  fit(training_data)

forecast(wf)
```

And comparing the result with the actual growth rates at that point in time,
```{r growth_rate_results}
growth_rates <- covid_case_death_rates |>
  filter(geo_value %in% used_locations) |>
  group_by(geo_value) |>
  mutate(
    # Multiply by 7 to estimate weekly equivalents
    case_gr = growth_rate(x = time_value, y = case_rate) * 7
  ) |>
  ungroup()

> growth_rates |> filter(time_value == "2021-08-01")

Reviewer 2: ```suggestion
growth_rates |> filter(time_value == "2021-08-01") |> select(-death_rate)
```
Since we're only looking at cases.
